Move [WKWebViewConfiguration _respectsImageOrientation] to API::PageConfiguration

squelart opened the pull request:
#### 19739c622738273c29292e7a1a65f18435981a97
<pre>
Move [WKWebViewConfiguration _respectsImageOrientation] to API::PageConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=270557">https://bugs.webkit.org/show_bug.cgi?id=270557</a>
<a href="https://rdar.apple.com/124115170">rdar://124115170</a>

Reviewed by Alex Christensen.

This flag already exists in WebPreferences, named &quot;shouldRespectImageOrientation&quot;,
so it&apos;s just a matter of using it directly through
`API::PageConfiguration::preferences()`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
The flag is already set when copying the WebPreferences pointer above.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
The flag is already initialized to the same platform-dependent value based on
Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml

(-[WKWebViewConfiguration copyWithZone:]):
The flag is already copied as part of _pageConfiguration&apos;s preferences above.

(-[WKWebViewConfiguration _respectsImageOrientation]):
(-[WKWebViewConfiguration _setRespectsImageOrientation:]):
Directly access the flag in _pageConfiguration&apos;s preferences through
self.preferences, which ensures it&apos;s lazily created if needed.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/cocoa/TestWKWebViewConfiguration.mm: Added.
(TEST):

Canonical link: <a href="https://commits.webkit.org/275929@main">https://commits.webkit.org/275929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cef948313a41edd88e5976b12aa0e1e72ec1e948

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43209 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22233 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45840 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39332 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45514 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19656 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35732 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37243 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16725 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38295 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1263 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39409 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38616 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47386 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18123 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42522 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19660 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41182 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9637 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19292 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -741,12 +734,12 @@ - (void)_setGroupIdentifier:(NSString *)groupIdentifier
 - (BOOL)_respectsImageOrientation
 {
-    return _respectsImageOrientation;
+    return _pageConfiguration->preferences()->shouldRespectImageOrientation();
```
I think preferences can be null, and this would introduce a crash.  In WKWebView._setupPageConfiguration we have already ensured that the preferences is not null, but that is not the case here.

squelart replied:
I assumed it was ok because of the other seemingly-unchecked access below, but in fact that access through `self.preferences` goes through a special getter that allocates a WebPreferences object if not there yet!
Ok, I'll see what I can do... I'd prefer not to have a redundant value in both PageConfiguration and WebPreferences -- but please let me know if you disagree.

achristensen07 commented on Tools/TestWebKitAPI/cocoa/TestWKWebViewConfiguration.mm:
```diff
@@ -0,0 +1,65 @@
+        BOOL defaultRespectsImageOrientation = [configuration _respectsImageOrientation];
+        auto view = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
+        EXPECT_EQ([[view configuration] _respectsImageOrientation], defaultRespectsImageOrientation);
+        // FIXME: Spooky action at a distance, due to API::PageConfiguration::copy() doing a shallow copy and keeping a reference to the same WebPreferences. Is that shallow copy intended?
```
It is in this case.

achristensen07 commented on Tools/TestWebKitAPI/cocoa/TestWKWebViewConfiguration.mm:
```diff
@@ -0,0 +1,65 @@
+#import <WebKit/WKWebViewConfigurationPrivate.h>
+#import <WebKit/WKWebViewPrivateForTesting.h>
+
+TEST(WKWebViewConfiguration, flags)
```
Flags with capital F to match the other tests.

squelart committed:
Move [WKWebViewConfiguration _respectsImageOrientation] to API::PageConfiguration
https://bugs.webkit.org/show_bug.cgi?id=270557
rdar://124115170

Reviewed by Alex Christensen.

This flag already exists in WebPreferences, named "shouldRespectImageOrientation",
so it's just a matter of using it directly through
`API::PageConfiguration::preferences()`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
The flag is already set when copying the WebPreferences pointer above.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
The flag is already initialized to the same platform-dependent value based on
Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml

(-[WKWebViewConfiguration copyWithZone:]):
The flag is already copied as part of _pageConfiguration's preferences above.

(-[WKWebViewConfiguration _respectsImageOrientation]):
(-[WKWebViewConfiguration _setRespectsImageOrientation:]):
Directly access the flag in _pageConfiguration's preferences through
self.preferences, which ensures it's lazily created if needed.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/cocoa/TestWKWebViewConfiguration.mm: Added.
(TEST):

Canonical link: https://commits.webkit.org/275929@main
